src/iop/demosaic.c

```diff
@@ -2422,6 +2426,452 @@ static void passthrough_monochrome(float *out, const float *const in, dt_iop_roi
+      const float W_Est = cfa[indx -  1] * (1.f + (lpf[indx] - lpf[indx -  2]) / ( eps + lpf[indx] + lpf[indx -  2]));
+      const float E_Est = cfa[indx +  1] * (1.f + (lpf[indx] - lpf[indx +  2]) / ( eps + lpf[indx] + lpf[indx +  2]));
+
+      // Vertical and horizontal estimations
+      const float V_Est = (S_Grad * N_Est + N_Grad * S_Est) / (N_Grad + S_Grad);
+      const float H_Est = (W_Grad * E_Est + E_Grad * W_Est) / (E_Grad + W_Grad);
+
+      // G@B and G@R interpolation
+      rgb[indx][1] = CLAMPS(VH_Disc * H_Est + (1.f - VH_Disc) * V_Est, 0.f, 1.f);
+    }
+  }
+
+  // STEP 4: Populate the red and blue channels
+  // Step 4.1: Calculate P/Q diagonal local discrimination
+#ifdef _OPENMP
+#pragma omp parallel for \
+  dt_omp_firstprivate(cfa, PQ_Dir, filters) \
+  schedule(static)
+#endif
+  for(int row = 4; row < height - 4; row++)
+  {
+    for(int col = 4 + (FC(row + roi_dy, roi_dx, filters) & 1); col < width - 4; col += 2)
+    {
+      const int indx = row * width + col;
+
+      const float P_Stat = fmaxf( -18.f * cfa[indx] * cfa[indx - w1 - 1] - 18.f * cfa[indx] * cfa[indx + w1 + 1] - 36.f * cfa[indx] * cfa[indx - w2 - 2] - 36.f * cfa[indx] * cfa[indx + w2 + 2] + 18.f * cfa[indx] * cfa[indx - w3 - 3] + 18.f * cfa[indx] * cfa[indx + w3 + 3] - 2.f * cfa[indx] * cfa[indx - w4 - 4] - 2.f * cfa[indx] * cfa[indx + w4 + 4] + 38.f * cfa[indx] * cfa[indx] - 70.f * cfa[indx - w1 - 1] * cfa[indx + w1 + 1] - 12.f * cfa[indx - w1 - 1] * cfa[indx - w2 - 2] + 24.f * cfa[indx - w1 - 1] * cfa[indx + w2 + 2] - 38.f * cfa[indx - w1 - 1] * cfa[indx - w3 - 3] + 16.f * cfa[indx - w1 - 1] * cfa[indx + w3 + 3] + 12.f * cfa[indx - w1 - 1] * cfa[indx - w4 - 4] - 6.f * cfa[indx - w1 - 1] * cfa[indx + w4 + 4] + 46.f * cfa[indx - w1 - 1] * cfa[indx - w1 - 1] + 24.f * cfa[indx + w1 + 1] * cfa[indx - w2 - 2] - 12.f * cfa[indx + w1 + 1] * cfa[indx + w2 + 2] + 16.f * cfa[indx + w1 + 1] * cfa[indx - w3 - 3] - 38.f * cfa[indx + w1 + 1] * cfa[indx + w3 + 3] - 6.f * cfa[indx + w1 + 1] * cfa[indx - w4 - 4] + 12.f * cfa[indx + w1 + 1] * cfa[indx + w4 + 4] + 46.f * cfa[indx + w1 + 1] * cfa[indx + w1 + 1] + 14.f * cfa[indx - w2 - 2] * cfa[indx + w2 + 2] - 12.f * cfa[indx - w2 - 2] * cfa[indx + w3 + 3] - 2.f * cfa[indx - w2 - 2] * cfa[indx - w4 - 4] + 2.f * cfa[indx - w2 - 2] * cfa[indx + w4 + 4] + 11.f * cfa[indx - w2 - 2] * cfa[indx - w2 - 2] - 12.f * cfa[indx + w2 + 2] * cfa[indx - w3 - 3] + 2 * cfa[indx + w2 + 2] * cfa[indx - w4 - 4] - 2.f * cfa[indx + w2 + 2] * cfa[indx + w4 + 4] + 11.f * cfa[indx + w2 + 2] * cfa[indx + w2 + 2] + 2.f * cfa[indx - w3 - 3] * cfa[indx + w3 + 3] - 6.f * cfa[indx - w3 - 3] * cfa[indx - w4 - 4] + 10.f * cfa[indx - w3 - 3] * cfa[indx - w3 - 3] - 6.f * cfa[indx + w3 + 3] * cfa[indx + w4 + 4] + 10.f * cfa[indx + w3 + 3] * cfa[indx + w3 + 3] + 1.f * cfa[indx - w4 - 4] * cfa[indx - w4 - 4] + 1.f * cfa[indx + w4 + 4] * cfa[indx + w4 + 4], epssq );
```

Comment:
any chance this could be formatted in a shape that makes some filter appear ? That's hardly legible.

Reply:
You mean like keeping line length in sensible order plus some formatting? Sure...

Reply:
should be better now

Reply:
I meant a way where the ordering of coefficients shows some logic. Something like : 
https://github.com/darktable-org/darktable/blob/bc2763e464277cd618be4b954dca40e65fd92113/src/iop/filmicrgb.c#L2092-L2097

But I understand there is no paper that explains what happens in the algo ?

Reply:
Unfortunately not. Some discussion about this at https://discuss.pixls.us/t/diagonal-interpolation-correction-artifacts-with-amaze-demosaicing/3338/94